man/lprint-server.1
.\"
.\" lprint-server man page for LPrint, a Label Printer Utility
.\"
.\" Copyright © 2019-2026 by Michael R Sweet.
.\"
.\" Licensed under Apache License v2.0.  See the file "LICENSE" for more
.\" information.
.\"
.TH lprint-server 1 "LPrint" "2026-06-03" "Michael R Sweet"
.SH NAME
lprint-server \- start a label print server.
.SH SYNOPSIS
.B lprint
.B server
[
.B \-o
.I OPTION=VALUE
]
.SH DESCRIPTION
This command starts a label print server.
Normally a server is started automatically as needed, however you can run the server manually to listen on a public network address and/or enable different logging options.
.PP
Server options can be specified on the command-line using the
.B -\o
.I OPTION=VALUE
arguments or in the
.B lprint.conf
configuration file.
.SH OPTIONS
The following options are recognized by
.B lprint server:
.TP 5
\fB\-o admin\-group=\fIGROUP\fR
Specifies that remote configuration is authenticated against the named group.
If not specified or the value "none" is given then any authenticated user may remotely configure lprint.
.TP 5
\fB\-o auth\-service=\fISERVICE\fR
Specifies the PAM service to use to authenticate for remote configuration requests.
If not specified or the value "none" is given then printers can only be added, modified, or deleted locally.
.TP 5
\fB\-o listen-hostname=\fIHOSTNAME\fR
Listens for IPP connections on the specified hostname/address(es).
If not specified, uses the wildcard addresses to allow connections from any address.
.TP 5
.B \-o log\-file=\-
Specifies that all messages are written to the standard error.
.TP 5
.B \-o log\-file=syslog
Specifies that all messages are sent to the syslog service.
.TP 5
\fB\-o log\-file=\fRFILENAME\fR
Specifies that all messages are written to the named file.
If the file already exists, messages are appended to it.
The default log file is "$TMPDIR/lprintUID.log".
.TP 5
.B \-o log\-level=debug
Specifies that all messages should be logged.
.TP 5
.B \-o log\-level=error
Specifies that fatal and error messages should be logged.
This is the default.
.TP 5
.B \-o log\-level=fatal
Specifies that only fatal messages should be logged.
.TP 5
.B \-o log\-level=info
Specifies that error, warning, and informational messages should be logged.
.TP 5
.B \-o log\-level=warning
Specifies that fatal, error, and warning messages should be logged.
.TP 5
\fB\-o server\-hostname=\fIHOSTNAME\fR
Specifies the hostname for the server.
The default is the system hostname.
.TP 5
\fB\-o server\-options=\fIOPTIONS\fI
Specifies a comma-delimited list of server options.
See "SERVER OPTIONS" below.
.TP 5
\fB\-o server\-port=\fIPORT\fR
Listens for IPP connections on the specified port.
If not specified, a random port between 8000 and 8999 is chosen.
.TP 5
\fB\-o spool\-directory=\fIDIRECTORY\fR
Specifies a directory that holds pending print files.
If not specified, a subdirectory in the system temporary directory is used.
.TP 5
\fB\-o system\-name=\fINAME\fR
Specifies the DNS-SD service name for the server.
The default is "LPrint".
.SH SERVER OPTIONS
By default,
.B lprint server
uses the "multi-queue", "web-interface", "web-log", "web-security", and "web-tls" options.
The \fB\-o server\-options=\fIOPTIONS\fR option can be used to override the default options, where "OPTIONS" is a comma-delimited list of any of the following:
.TP 5
.B dnssd\-host
When resolving conflicts, add the DNS-SD hostname instead of a serial number or UUID.
.TP 5
.B no\-multi\-queue
Do no support multiple print queues.
.TP 5
.B raw\-socket
Provide a raw ("JetDirect") socket for each printer on port 91xx.
.TP 5
.B usb\-printer
Enable a USB printer gadget for the default printer (Linux only for Raspberry Pi and similar embedded SBCs).
.TP 5
.B no\-web\-interface
Disables the web interface.
.TP 5
.B web\-log
Enables the log file web page.
.TP 5
.B web\-network
Enables the network settings web page.
.TP 5
.B web\-remote
Enables remote queue management.
.TP 5
.B web\-security
Enables the user/password settings web page.
.TP 5
.B no\-tls
Disables TLS support entirely.
.SH SEE ALSO
.BR lprint (1),
.BR lprint-shutdown (1),
.BR lprint.conf (5).
.SH COPYRIGHT
Copyright \[co] 2019-2026 by Michael R Sweet.


CHANGES.md
LPrint Change History
=====================

v1.4.1 - YYYY-MM-DD
-------------------

- Fixed build error when compiling against older versions of libcups
  (Issue #210)


v1.4.0 - 2026-06-08
-------------------

- Added ESC/POS driver (Issue #123)
- Added Vevor Y428BT driver (Issue #130)
- Added 4 x 7.83 inch label size (Issue #144)
- Added Arkscan 2054a and HP KE103 (Issue #165)
- Added DYMO LabelWriter Twin Turbo driver (Issue #193)
- Now enable the TLS web interface unless the "no-tls" option is specified
  (Issue #161)
- Updated the default state file to match current PAPPL defaults (Issue #129)
- Updated the maximum label width of TSPL printers to 105mm (Issue #141)
- Updated the ZPL driver to automatically disable status commands if the target
  printer doesn't implement that part of ZPL (Issue #161)
- Fixed the margins for DYMO printers (Issue #132)
- Fixed some Brother driver problems (Issue #167)
- Fixed a NumCopies issue in the TSPL driver (Issue #188)
- Fixed list of label sizes for DYMO LabelWriter 4XL (Issue #198)


v1.3.1 - 2024-02-09
-------------------

- Updated "print-speed" support in TSPL driver (Issue #120 and #121)
- Fixed lprint-modify man page (Issue #122 and #126)
- Fixed snap documentation for connecting LPrint to Avahi.


v1.3.0 - 2024-01-31
-------------------

- Added new dithering algorithm to better support barcode printing with shaded
  content.
- Added experimental Brother printer support (Issue #15)
- Added basic TSPL printer support (Issue #54)
- Added basic SEIKO printer support (Issue #58)
- Added experimental Zebra CPCL printer support.
- Added support for configuration files in "/etc", "/usr/local/etc", or
  "/Library/Application Support" (macOS).
- Updated ZPL driver to report 'media-needed' reason when out of labels during a
  job.
- Fixed copies support for ZPL printers (Issue #100)
- Fixed darkness calculations for EPL and ZPL printers (Issue #104)


v1.2.0 - 2022-12-31
-------------------

- Documentation corrections (Issue #53, Issue #76)
- Added snap server configuration.
- Added `--with-systemd` configure option and install to
 `$prefix/lib/systemd/system` by default (Issue #50)
- Added EPL2 and ZPL auto-typing support (Issue #64)
- Changed the default log level for systemd to info (Issue #82)
- Fixed macOS installer to start LPrint server (Issue #48)
- Fixed configure script not using warning/preprocessor options in build
  (Issue #60)
- Fixed printer renaming algorithm to not truncate the name (Issue #60)
- Fixed missing "print-color-mode=bi-level" for bar code printing (Issue #76)
- Fixed label mode support in EPL and ZPL drivers (Issue #79)
- Fixed driver names for EPL printers (Issue #52)


v1.1.0 - 2021-12-23
-------------------

- Switched to PAPPL (Issue #20, #35)
- Fixed `lprint default` and `lprint delete` not working (Issue #17, Issue #40)
- Fixed server crashes on `SIGINT` (Issue #18)
- Fixed the reported date and time information when no printers were added
  (Issue #26)
- Fixed a startup bug on macOS (Issue #34)
- Now support auto-detection of printer drivers and auto-add USB printers the
  first time LPrint is run.
- Added the missing macOS binary package (Issue #33)
- Added launchd and systemd service files for running lprint as a service.
- Added support for Zebra/Eltron EPL2 printers (Issue #38)
- Added support for 600dpi ZPL thermal transfer printers.
- Added support for ZPL status/ready media updates.
- Added support for test pages.
- Temporarily removed support for DYMO LabelWriter Wireless printer (Issue #23)


v1.0 - 2020-02-17
-----------------

- First production release.


v1.0rc1 - 2020-02-10
--------------------

- Updated media documentation (Issue #13)
- The `lprint options` command now reports both English and metric dimensions
  for all supported sizes (Issue #14)
- Now support setting the default media source, top offset, tracking, and type
  values.
- Now save and load next job-id value for each printer, along with the
  printer-config-change-[date-]time and printer-impressions-completed values.
- Fixed some small memory leaks.


v1.0b2 - 2020-01-26
-------------------

- Added support for authentication of remote administrative requests (Issue #1)
- Added support for managing printers via web browser (Issue #1)
- The `add`, `default`, `delete`, and `modify` sub-commands now support the "-u"
  option (Issue #1)
- Fixed an ASLR issue with some Linux compilers (Issue #5)
- Added code to detach the Linux usblp kernel driver since apparently there are
  still Linux distributions shipping the old (and broken) usblp kernel driver
  (Issue #7)
- Device errors are now logged to stderr (for `lprint devices`) or the server
  log (for `lprint server`) so that it is possible to see permission and other
  access errors (Issue #8)
- Job attributes are now validated properly against the driver attributes
  (Issue #9)
- Fixed an issue in server mode - a failed IPv6 listener would prevent the
  server from starting (Issue #12)
- Added a "spool-directory" option for the `server` sub-command.
- Added a spec file for building RPMs.


v1.0b1 - 2020-01-15
-------------------

- First public release.


lprint-server.c
//
// Server sub-command for LPrint, a Label Printer Application
//
// Copyright © 2019 by Michael R Sweet.
//
// Licensed under Apache License v2.0.  See the file "LICENSE" for more
// information.
//

//
// Include necessary headers...
//

#include "lprint.h"
#include <ctype.h>


//
// 'lprintDoServer()' - Do the server sub-command.
//

int					// O - Exit status
lprintDoServer(
    int           num_options,		// I - Number of options
    cups_option_t *options)		// I - Options
{
  lprint_system_t *system;		// System object
  const char	*val;			// Option value
  lprint_loglevel_t loglevel;		// Log level
  const char	*logfile,		// Log file, if any
		*hostname;		// Hostname, if any
  int		port;			// Port number, if any


  // Get config options...
  if ((val = cupsGetOption("log-level", num_options, options)) != NULL)
  {
    if (!strcmp(val, "fatal"))
      loglevel = LPRINT_LOGLEVEL_FATAL;
    else if (!strcmp(val, "error"))
      loglevel = LPRINT_LOGLEVEL_ERROR;
    else if (!strcmp(val, "warn"))
      loglevel = LPRINT_LOGLEVEL_WARN;
    else if (!strcmp(val, "info"))
      loglevel = LPRINT_LOGLEVEL_INFO;
    else if (!strcmp(val, "debug"))
      loglevel = LPRINT_LOGLEVEL_DEBUG;
    else
    {
      fprintf(stderr, "lprint: Bad log-level value '%s'.\n", val);
      return (1);
    }
  }
  else
    loglevel = LPRINT_LOGLEVEL_UNSPEC;

  logfile  = cupsGetOption("log-file", num_options, options);
  hostname = cupsGetOption("server-name", num_options, options);

  if ((val = cupsGetOption("server-port", num_options, options)) != NULL)
  {
    if (!isdigit(*val & 255))
    {
      fprintf(stderr, "lprint: Bad server-port value '%s'.\n", val);
      return (1);
    }
    else
      port = atoi(val);
  }
  else
    port = 0;

  // Create the system object and run it...
  if ((system = lprintCreateSystem(hostname, port, hostname ? "_print,_universal" : NULL, cupsGetOption("spool-directory", num_options, options), logfile, loglevel, cupsGetOption("auth-service", num_options, options), cupsGetOption("admin-group", num_options, options))) == NULL)
    return (1);

  lprintRunSystem(system);
  lprintDeleteSystem(system);

  return (0);
}
